sql/parser/eval.go

```diff
@@ -841,6 +856,45 @@ func evalFuncExpr(expr *FuncExpr) (Datum, error) {
 }
 
 func evalCaseExpr(expr *CaseExpr) (Datum, error) {
+	whenVals := make([]Datum, 0, len(expr.Whens))
+	var condType, valType reflect.Type
+	for _, when := range expr.Whens {
+		cond, err := EvalExpr(when.Cond)
+		if err != nil {
+			return DNull, err
+		}
+		if condType == nil {
+			condType = reflect.TypeOf(cond)
+		} else if cond != DNull && reflect.TypeOf(cond) != condType {
+			return DNull, fmt.Errorf("incompatible condition type %s", cond.Type())
+		}
+
+		val, err := EvalExpr(when.Val)
+		if err != nil {
+			return DNull, err
+		}
+
+		if valType == nil {
+			valType = reflect.TypeOf(val)
+		} else if val != DNull && reflect.TypeOf(val) != valType {
+			return DNull, fmt.Errorf("incompatible value type %s", val.Type())
+		}
+
+		whenVals = append(whenVals, val)
+	}
+
+	var elseVal Datum
+	if expr.Else != nil {
+		val, err := EvalExpr(expr.Else)
+		if err != nil {
+			return DNull, err
+		}
+		if val != DNull && reflect.TypeOf(val) != valType {
+			return DNull, fmt.Errorf("incompatible value type %s", val.Type())
+		}
+		elseVal = val
+	}
```

Comment:
So I added the test case `SELECT ... WHERE CASE WHEN x <> 0 THEN y/x > 1.5 ELSE false END`, and it seems to work, which is the case because EvalExpr isn't substituting the column values. I'm not sure I understand the `NormalizeExpr` suggestion since that returns an `Expr` and we need a `Datum`.